<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>An Introduction to Machine Learning</title>
  <meta name="description" content="Course materials for An Introduction to Machine Learning">
  <meta name="generator" content="bookdown 0.5 and GitBook 2.6.7">

  <meta property="og:title" content="An Introduction to Machine Learning" />
  <meta property="og:type" content="book" />
  
  <meta property="og:image" content="figures/cover_image.png" />
  <meta property="og:description" content="Course materials for An Introduction to Machine Learning" />
  <meta name="github-repo" content="bioinformatics-training/intro-machine-learning" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="An Introduction to Machine Learning" />
  
  <meta name="twitter:description" content="Course materials for An Introduction to Machine Learning" />
  <meta name="twitter:image" content="figures/cover_image.png" />

<meta name="author" content="Sudhakaran Prabakaran, Matt Wayland and Chris Penfold">


<meta name="date" content="2017-08-30">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="dimensionality-reduction.html">
<link rel="next" href="resources.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">An Introduction to Machine Learning</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About the course</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#overview"><i class="fa fa-check"></i><b>1.1</b> Overview</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#registration"><i class="fa fa-check"></i><b>1.2</b> Registration</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.3</b> Prerequisites</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#github"><i class="fa fa-check"></i><b>1.4</b> Github</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.5</b> License</a></li>
<li class="chapter" data-level="1.6" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.6</b> Contact</a></li>
<li class="chapter" data-level="1.7" data-path="index.html"><a href="index.html#colophon"><i class="fa fa-check"></i><b>1.7</b> Colophon</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a></li>
<li class="chapter" data-level="3" data-path="linear-models-and-matrix-algebra.html"><a href="linear-models-and-matrix-algebra.html"><i class="fa fa-check"></i><b>3</b> Linear models and matrix algebra</a></li>
<li class="chapter" data-level="4" data-path="linear-and-non-linear-logistic-regression.html"><a href="linear-and-non-linear-logistic-regression.html"><i class="fa fa-check"></i><b>4</b> Linear and non linear logistic regression</a></li>
<li class="chapter" data-level="5" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html"><i class="fa fa-check"></i><b>5</b> Nearest neighbours</a><ul>
<li class="chapter" data-level="5.1" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#example-one"><i class="fa fa-check"></i><b>5.1</b> Example one</a></li>
<li class="chapter" data-level="5.2" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#example-two"><i class="fa fa-check"></i><b>5.2</b> Example two</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="decision-trees-and-random-forests.html"><a href="decision-trees-and-random-forests.html"><i class="fa fa-check"></i><b>6</b> Decision trees and random forests</a></li>
<li class="chapter" data-level="7" data-path="support-vector-machines.html"><a href="support-vector-machines.html"><i class="fa fa-check"></i><b>7</b> Support vector machines</a></li>
<li class="chapter" data-level="8" data-path="artificial-neural-networks.html"><a href="artificial-neural-networks.html"><i class="fa fa-check"></i><b>8</b> Artificial neural networks</a></li>
<li class="chapter" data-level="9" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html"><i class="fa fa-check"></i><b>9</b> Dimensionality reduction</a><ul>
<li class="chapter" data-level="9.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#linear-dimensionality-reduction"><i class="fa fa-check"></i><b>9.1</b> Linear Dimensionality Reduction</a><ul>
<li class="chapter" data-level="9.1.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#principle-component-analysis"><i class="fa fa-check"></i><b>9.1.1</b> Principle Component Analysis</a></li>
<li class="chapter" data-level="9.1.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#horeshoe-effect"><i class="fa fa-check"></i><b>9.1.2</b> Horeshoe effect</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#nonlinear-dimensionality-reduction"><i class="fa fa-check"></i><b>9.2</b> Nonlinear Dimensionality Reduction</a><ul>
<li class="chapter" data-level="9.2.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#t-sne"><i class="fa fa-check"></i><b>9.2.1</b> t-SNE</a></li>
<li class="chapter" data-level="9.2.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#gaussian-process-latent-variable-models"><i class="fa fa-check"></i><b>9.2.2</b> Gaussian Process Latent Variable Models</a></li>
<li class="chapter" data-level="9.2.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#gplvms-with-informative-priors"><i class="fa fa-check"></i><b>9.2.3</b> GPLVMs with informative priors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>10</b> Clustering</a><ul>
<li class="chapter" data-level="10.1" data-path="clustering.html"><a href="clustering.html#introduction"><i class="fa fa-check"></i><b>10.1</b> Introduction</a></li>
<li class="chapter" data-level="10.2" data-path="clustering.html"><a href="clustering.html#distance-metrics"><i class="fa fa-check"></i><b>10.2</b> Distance metrics</a><ul>
<li class="chapter" data-level="10.2.1" data-path="clustering.html"><a href="clustering.html#image-segmentation"><i class="fa fa-check"></i><b>10.2.1</b> Image segmentation</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="clustering.html"><a href="clustering.html#hierarchic-methods"><i class="fa fa-check"></i><b>10.3</b> Hierarchic methods</a><ul>
<li class="chapter" data-level="10.3.1" data-path="clustering.html"><a href="clustering.html#linkage-algorithms"><i class="fa fa-check"></i><b>10.3.1</b> Linkage algorithms</a></li>
<li class="chapter" data-level="10.3.2" data-path="clustering.html"><a href="clustering.html#example-clustering-toy-data-sets"><i class="fa fa-check"></i><b>10.3.2</b> Example: clustering toy data sets</a></li>
<li class="chapter" data-level="10.3.3" data-path="clustering.html"><a href="clustering.html#example-gene-expression-profiling-of-human-tissues"><i class="fa fa-check"></i><b>10.3.3</b> Example: gene expression profiling of human tissues</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="clustering.html"><a href="clustering.html#partitioning-methods"><i class="fa fa-check"></i><b>10.4</b> Partitioning methods</a><ul>
<li class="chapter" data-level="10.4.1" data-path="clustering.html"><a href="clustering.html#k-means"><i class="fa fa-check"></i><b>10.4.1</b> K-means</a></li>
<li class="chapter" data-level="10.4.2" data-path="clustering.html"><a href="clustering.html#dbscan"><i class="fa fa-check"></i><b>10.4.2</b> DBSCAN</a></li>
<li class="chapter" data-level="10.4.3" data-path="clustering.html"><a href="clustering.html#gene-expression"><i class="fa fa-check"></i><b>10.4.3</b> Gene expression</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="clustering.html"><a href="clustering.html#summary"><i class="fa fa-check"></i><b>10.5</b> Summary</a><ul>
<li class="chapter" data-level="10.5.1" data-path="clustering.html"><a href="clustering.html#applications"><i class="fa fa-check"></i><b>10.5.1</b> Applications</a></li>
<li class="chapter" data-level="10.5.2" data-path="clustering.html"><a href="clustering.html#strengths"><i class="fa fa-check"></i><b>10.5.2</b> Strengths</a></li>
<li class="chapter" data-level="10.5.3" data-path="clustering.html"><a href="clustering.html#limitations"><i class="fa fa-check"></i><b>10.5.3</b> Limitations</a></li>
</ul></li>
<li class="chapter" data-level="10.6" data-path="clustering.html"><a href="clustering.html#exercises"><i class="fa fa-check"></i><b>10.6</b> Exercises</a></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>A</b> Resources</a><ul>
<li class="chapter" data-level="A.1" data-path="resources.html"><a href="resources.html#python"><i class="fa fa-check"></i><b>A.1</b> Python</a></li>
<li class="chapter" data-level="A.2" data-path="resources.html"><a href="resources.html#machine-learning-data-set-repository"><i class="fa fa-check"></i><b>A.2</b> Machine learning data set repository</a></li>
</ul></li>
<li class="chapter" data-level="B" data-path="solutions-to-exercises.html"><a href="solutions-to-exercises.html"><i class="fa fa-check"></i><b>B</b> Solutions to exercises</a><ul>
<li class="chapter" data-level="B.1" data-path="solutions-to-exercises.html"><a href="solutions-to-exercises.html#solutions-linear-models"><i class="fa fa-check"></i><b>B.1</b> Chapter 2 - Linear models and matrix algebra</a></li>
<li class="chapter" data-level="B.2" data-path="solutions-to-exercises.html"><a href="solutions-to-exercises.html#solutions-logistic-regression"><i class="fa fa-check"></i><b>B.2</b> Chapter 3 - Linear and non-linear logistic regression</a></li>
<li class="chapter" data-level="B.3" data-path="solutions-to-exercises.html"><a href="solutions-to-exercises.html#solutions-nearest-neighbours"><i class="fa fa-check"></i><b>B.3</b> Chapter 4 - Nearest neighbours</a></li>
<li class="chapter" data-level="B.4" data-path="solutions-to-exercises.html"><a href="solutions-to-exercises.html#solutions-decision-trees"><i class="fa fa-check"></i><b>B.4</b> Chapter 5 - Decision trees and random forests</a></li>
<li class="chapter" data-level="B.5" data-path="solutions-to-exercises.html"><a href="solutions-to-exercises.html#solutions-svm"><i class="fa fa-check"></i><b>B.5</b> Chapter 6 - Support vector machines</a></li>
<li class="chapter" data-level="B.6" data-path="solutions-to-exercises.html"><a href="solutions-to-exercises.html#solutions-ann"><i class="fa fa-check"></i><b>B.6</b> Chapter 7 - Artificial neural networks</a></li>
<li class="chapter" data-level="B.7" data-path="solutions-to-exercises.html"><a href="solutions-to-exercises.html#solutions-dimensionality-reduction"><i class="fa fa-check"></i><b>B.7</b> Chapter 8 - Dimensionality reduction</a></li>
<li class="chapter" data-level="B.8" data-path="solutions-to-exercises.html"><a href="solutions-to-exercises.html#solutions-clustering"><i class="fa fa-check"></i><b>B.8</b> Chapter 9 - Clustering</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">An Introduction to Machine Learning</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="clustering" class="section level1">
<h1><span class="header-section-number">10</span> Clustering</h1>
<!-- Matt -->
<!-- http://www.madgroup.path.cam.ac.uk/microarraysummary.shtml -->
<div id="introduction" class="section level2">
<h2><span class="header-section-number">10.1</span> Introduction</h2>
<p>Hierarchic (produce dendrogram) vs partitioning methods</p>
<div class="figure" style="text-align: center"><span id="fig:clusterTypes"></span>
<img src="09-clustering_files/figure-html/clusterTypes-1.png" alt="Example clusters. **A**, *blobs*; **B**, *aggregation* [@Gionis2007]; **C**, *noisy moons*; **D**, *noisy circles*; **E**, *D31* [@Veenman2002]; **F**, *no structure*." width="80%" />
<p class="caption">
Figure 10.1: Example clusters. <strong>A</strong>, <em>blobs</em>; <strong>B</strong>, <em>aggregation</em> <span class="citation">(Gionis, Mannila, and Tsaparas <a href="#ref-Gionis2007">2007</a>)</span>; <strong>C</strong>, <em>noisy moons</em>; <strong>D</strong>, <em>noisy circles</em>; <strong>E</strong>, <em>D31</em> <span class="citation">(Veenman, Reinders, and Backer <a href="#ref-Veenman2002">2002</a>)</span>; <strong>F</strong>, <em>no structure</em>.
</p>
</div>
</div>
<div id="distance-metrics" class="section level2">
<h2><span class="header-section-number">10.2</span> Distance metrics</h2>
<strong>Minkowski distance:</strong>
<span class="math display" id="eq:minkowski">\[\begin{equation}
  distance\left(x,y,p\right)=\left(\sum_{i=1}^{n} abs(x_i-y_i)^p\right)^{1/p}
  \tag{10.1}
\end{equation}\]</span>
<p>Graphical explanation of euclidean, manhattan and max (Chebyshev?)</p>
<div id="image-segmentation" class="section level3">
<h3><span class="header-section-number">10.2.1</span> Image segmentation</h3>
</div>
</div>
<div id="hierarchic-methods" class="section level2">
<h2><span class="header-section-number">10.3</span> Hierarchic methods</h2>
<table>
<caption><span id="tab:distance-matrix">Table 10.1: </span>Example distance matrix</caption>
<thead>
<tr class="header">
<th></th>
<th align="left">A</th>
<th align="left">B</th>
<th align="left">C</th>
<th align="left">D</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>B</td>
<td align="left">2</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="even">
<td>C</td>
<td align="left">6</td>
<td align="left">5</td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="odd">
<td>D</td>
<td align="left">10</td>
<td align="left">10</td>
<td align="left">5</td>
<td align="left"></td>
</tr>
<tr class="even">
<td>E</td>
<td align="left">9</td>
<td align="left">8</td>
<td align="left">3</td>
<td align="left">4</td>
</tr>
</tbody>
</table>
<div id="linkage-algorithms" class="section level3">
<h3><span class="header-section-number">10.3.1</span> Linkage algorithms</h3>
<p>Make one section panel of three dendrograms one table</p>
<p>Single linkage - nearest neighbours linkage Complete linkage - furthest neighbours linkage Average linkage - UPGMA (Unweighted Pair Group Method with Arithmetic Mean)</p>
<table>
<caption><span id="tab:distance-merge">Table 10.2: </span>Merge distances for objects in the example distance matrix using three different linkage methods.</caption>
<thead>
<tr class="header">
<th align="left">Groups</th>
<th align="left">Single</th>
<th align="left">Complete</th>
<th align="left">Average</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">A,B,C,D,E</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
</tr>
<tr class="even">
<td align="left">(A,B),C,D,E</td>
<td align="left">2</td>
<td align="left">2</td>
<td align="left">2</td>
</tr>
<tr class="odd">
<td align="left">(A,B),(C,E),D</td>
<td align="left">3</td>
<td align="left">3</td>
<td align="left">3</td>
</tr>
<tr class="even">
<td align="left">(A,B)(C,D,E)</td>
<td align="left">4</td>
<td align="left">5</td>
<td align="left">4.5</td>
</tr>
<tr class="odd">
<td align="left">(A,B,C,D,E)</td>
<td align="left">5</td>
<td align="left">10</td>
<td align="left">8</td>
</tr>
</tbody>
</table>
<div class="figure" style="text-align: center"><span id="fig:linkageComparison"></span>
<img src="09-clustering_files/figure-html/linkageComparison-1.png" alt="Dendrograms for the example distance matrix using three different linkage methods. " width="100%" /><img src="09-clustering_files/figure-html/linkageComparison-2.png" alt="Dendrograms for the example distance matrix using three different linkage methods. " width="100%" /><img src="09-clustering_files/figure-html/linkageComparison-3.png" alt="Dendrograms for the example distance matrix using three different linkage methods. " width="100%" />
<p class="caption">
Figure 10.2: Dendrograms for the example distance matrix using three different linkage methods.
</p>
</div>
</div>
<div id="example-clustering-toy-data-sets" class="section level3">
<h3><span class="header-section-number">10.3.2</span> Example: clustering toy data sets</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RColorBrewer)
<span class="kw">library</span>(dendextend)</code></pre></div>
<pre><code>## 
## ---------------------
## Welcome to dendextend version 1.5.2
## Type citation(&#39;dendextend&#39;) for how to cite the package.
## 
## Type browseVignettes(package = &#39;dendextend&#39;) for the package vignette.
## The github page is: https://github.com/talgalili/dendextend/
## 
## Suggestions and bug-reports can be submitted at: https://github.com/talgalili/dendextend/issues
## Or contact: &lt;tal.galili@gmail.com&gt;
## 
##  To suppress this message use:  suppressPackageStartupMessages(library(dendextend))
## ---------------------</code></pre>
<pre><code>## 
## Attaching package: &#39;dendextend&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:ggdendro&#39;:
## 
##     theme_dendro</code></pre>
<pre><code>## The following object is masked from &#39;package:stats&#39;:
## 
##     cutree</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
<span class="kw">library</span>(GGally)

cluster_colours &lt;-<span class="st"> </span><span class="kw">brewer.pal</span>(<span class="dv">8</span>,<span class="st">&quot;Dark2&quot;</span>)

blobs &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;data/example_clusters/blobs.csv&quot;</span>, <span class="dt">header=</span>F)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">aggregation &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/example_clusters/aggregation.txt&quot;</span>)
noisy_moons &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;data/example_clusters/noisy_moons.csv&quot;</span>, <span class="dt">header=</span>F)
noisy_circles &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;data/example_clusters/noisy_circles.csv&quot;</span>, <span class="dt">header=</span>F)
no_structure &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;data/example_clusters/no_structure.csv&quot;</span>, <span class="dt">header=</span>F)

hclust_plots &lt;-<span class="st"> </span>function(data_set, n){
  d &lt;-<span class="st"> </span><span class="kw">dist</span>(data_set[,<span class="dv">1</span>:<span class="dv">2</span>])
  dend &lt;-<span class="st"> </span><span class="kw">as.dendrogram</span>(<span class="kw">hclust</span>(d, <span class="dt">method=</span><span class="st">&quot;average&quot;</span>))
  clusters &lt;-<span class="st"> </span><span class="kw">cutree</span>(dend,n,<span class="dt">order_clusters_as_data=</span>F)
  dend &lt;-<span class="st"> </span><span class="kw">color_branches</span>(dend, <span class="dt">clusters=</span>clusters, <span class="dt">col=</span>cluster_colours[<span class="dv">1</span>:n])
  clusters &lt;-<span class="st"> </span>clusters[<span class="kw">order</span>(<span class="kw">as.numeric</span>(<span class="kw">names</span>(clusters)))]
  <span class="kw">labels</span>(dend) &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="st">&quot;&quot;</span>, <span class="kw">length</span>(data_set[,<span class="dv">1</span>]))
  ggd &lt;-<span class="st"> </span><span class="kw">as.ggdend</span>(dend)
  ggd$nodes &lt;-<span class="st"> </span>ggd$nodes[!(<span class="dv">1</span>:<span class="kw">length</span>(ggd$nodes[,<span class="dv">1</span>])),]
  plotPair &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="kw">ggplot</span>(ggd),
    <span class="kw">ggplot</span>(data_set, <span class="kw">aes</span>(V1,V2)) +<span class="st"> </span><span class="kw">geom_point</span>(<span class="dt">col=</span>cluster_colours[clusters], <span class="dt">size=</span><span class="fl">0.2</span>))
  <span class="kw">return</span>(plotPair)
}

plotList &lt;-<span class="st"> </span><span class="kw">c</span>(
  <span class="kw">hclust_plots</span>(aggregation, <span class="dv">7</span>),
  <span class="kw">hclust_plots</span>(noisy_moons, <span class="dv">2</span>),
  <span class="kw">hclust_plots</span>(noisy_circles, <span class="dv">2</span>),
  <span class="kw">hclust_plots</span>(no_structure, <span class="dv">3</span>)
)

pm &lt;-<span class="st"> </span><span class="kw">ggmatrix</span>(
  plotList, <span class="dt">nrow=</span><span class="dv">4</span>, <span class="dt">ncol=</span><span class="dv">2</span>, <span class="dt">showXAxisPlotLabels =</span> F, <span class="dt">showYAxisPlotLabels =</span> F
) +<span class="st"> </span><span class="kw">theme_bw</span>()

pm</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:hclustToyData"></span>
<img src="09-clustering_files/figure-html/hclustToyData-1.png" alt="Hierarchical clustering of toy data-sets. " width="80%" />
<p class="caption">
Figure 10.3: Hierarchical clustering of toy data-sets.
</p>
</div>
</div>
<div id="example-gene-expression-profiling-of-human-tissues" class="section level3">
<h3><span class="header-section-number">10.3.3</span> Example: gene expression profiling of human tissues</h3>
<p>Load required libraries</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RColorBrewer)
<span class="kw">library</span>(dendextend)</code></pre></div>
<p>Load data</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&quot;data/tissues_gene_expression/tissuesGeneExpression.rda&quot;</span>)</code></pre></div>
<p>Inspect data</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(tissue)</code></pre></div>
<pre><code>## tissue
##  cerebellum       colon endometrium hippocampus      kidney       liver 
##          38          34          15          31          39          26 
##    placenta 
##           6</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(e)</code></pre></div>
<pre><code>## [1] 22215   189</code></pre>
<p>Compute distance between each sample</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">d &lt;-<span class="st"> </span><span class="kw">dist</span>(<span class="kw">t</span>(e))</code></pre></div>
<p>perform hierarchical clustering</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc &lt;-<span class="st"> </span><span class="kw">hclust</span>(d, <span class="dt">method=</span><span class="st">&quot;average&quot;</span>)
<span class="kw">plot</span>(hc, <span class="dt">labels=</span>tissue, <span class="dt">cex=</span><span class="fl">0.5</span>, <span class="dt">hang=</span>-<span class="dv">1</span>, <span class="dt">xlab=</span><span class="st">&quot;&quot;</span>, <span class="dt">sub=</span><span class="st">&quot;&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:tissueDendrogram"></span>
<img src="09-clustering_files/figure-html/tissueDendrogram-1.png" alt="Clustering of tissue samples based on gene expression profiles. " width="100%" />
<p class="caption">
Figure 10.4: Clustering of tissue samples based on gene expression profiles.
</p>
</div>
<p>use dendextend library to plot dendrogram with colour labels</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tissue_type &lt;-<span class="st"> </span><span class="kw">unique</span>(tissue)
dend &lt;-<span class="st"> </span><span class="kw">as.dendrogram</span>(hc)
dend_colours &lt;-<span class="st"> </span><span class="kw">brewer.pal</span>(<span class="kw">length</span>(<span class="kw">unique</span>(tissue)),<span class="st">&quot;Dark2&quot;</span>)
<span class="kw">names</span>(dend_colours) &lt;-<span class="st"> </span>tissue_type
<span class="kw">labels</span>(dend) &lt;-<span class="st"> </span>tissue[<span class="kw">order.dendrogram</span>(dend)]
<span class="kw">labels_colors</span>(dend) &lt;-<span class="st"> </span>dend_colours[tissue][<span class="kw">order.dendrogram</span>(dend)]
<span class="kw">labels_cex</span>(dend) =<span class="st"> </span><span class="fl">0.5</span>
<span class="kw">plot</span>(dend, <span class="dt">horiz=</span>T)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:tissueDendrogramColour"></span>
<img src="09-clustering_files/figure-html/tissueDendrogramColour-1.png" alt="Clustering of tissue samples based on gene expression profiles with labels coloured by tissue type. " width="100%" />
<p class="caption">
Figure 10.5: Clustering of tissue samples based on gene expression profiles with labels coloured by tissue type.
</p>
</div>
<p>Define clusters by cutting tree at a specific height</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(dend, <span class="dt">horiz=</span>T)
<span class="kw">abline</span>(<span class="dt">v=</span><span class="dv">125</span>, <span class="dt">lwd=</span><span class="dv">2</span>, <span class="dt">lty=</span><span class="dv">2</span>, <span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:tissueDendrogramCutHeight"></span>
<img src="09-clustering_files/figure-html/tissueDendrogramCutHeight-1.png" alt="Clusters found by cutting tree at a height of 125" width="100%" />
<p class="caption">
Figure 10.6: Clusters found by cutting tree at a height of 125
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hclusters &lt;-<span class="st"> </span><span class="kw">cutree</span>(dend, <span class="dt">h=</span><span class="dv">125</span>)
<span class="kw">table</span>(tissue, <span class="dt">cluster=</span>hclusters)</code></pre></div>
<pre><code>##              cluster
## tissue         1  2  3  4  5  6
##   cerebellum   0 36  0  0  2  0
##   colon        0  0 34  0  0  0
##   endometrium 15  0  0  0  0  0
##   hippocampus  0 31  0  0  0  0
##   kidney      37  0  0  0  2  0
##   liver        0  0  0 24  2  0
##   placenta     0  0  0  0  0  6</code></pre>
<p>Select a specific number of clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(dend, <span class="dt">horiz=</span>T)
<span class="kw">abline</span>(<span class="dt">v =</span> <span class="kw">heights_per_k.dendrogram</span>(dend)[<span class="st">&quot;8&quot;</span>], <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">lty =</span> <span class="dv">2</span>, <span class="dt">col =</span> <span class="st">&quot;blue&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:tissueDendrogramEightClusters"></span>
<img src="09-clustering_files/figure-html/tissueDendrogramEightClusters-1.png" alt="Selection of eight clusters from the dendogram" width="100%" />
<p class="caption">
Figure 10.7: Selection of eight clusters from the dendogram
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hclusters &lt;-<span class="st"> </span><span class="kw">cutree</span>(dend, <span class="dt">k=</span><span class="dv">8</span>)
<span class="kw">table</span>(tissue, <span class="dt">cluster=</span>hclusters)</code></pre></div>
<pre><code>##              cluster
## tissue         1  2  3  4  5  6  7  8
##   cerebellum   0 31  0  0  2  0  5  0
##   colon        0  0 34  0  0  0  0  0
##   endometrium  0  0  0  0  0 15  0  0
##   hippocampus  0 31  0  0  0  0  0  0
##   kidney      37  0  0  0  2  0  0  0
##   liver        0  0  0 24  2  0  0  0
##   placenta     0  0  0  0  0  0  0  6</code></pre>
</div>
</div>
<div id="partitioning-methods" class="section level2">
<h2><span class="header-section-number">10.4</span> Partitioning methods</h2>
<div id="k-means" class="section level3">
<h3><span class="header-section-number">10.4.1</span> K-means</h3>
<p>Pseudocode</p>
<p>to illustrate range of different types of data that can be clustered - image segmentation</p>
<div class="figure" style="text-align: center"><span id="fig:kmeansIterations"></span>
<img src="09-clustering_files/figure-html/kmeansIterations-1.png" alt="Iterations of the k-means algorithm" width="90%" />
<p class="caption">
Figure 10.8: Iterations of the k-means algorithm
</p>
</div>
</div>
<div id="dbscan" class="section level3">
<h3><span class="header-section-number">10.4.2</span> DBSCAN</h3>
<p>Density-based spatial clustering of applications with noise</p>
</div>
<div id="gene-expression" class="section level3">
<h3><span class="header-section-number">10.4.3</span> Gene expression</h3>
<p>tissue types?</p>
</div>
</div>
<div id="summary" class="section level2">
<h2><span class="header-section-number">10.5</span> Summary</h2>
<div id="applications" class="section level3">
<h3><span class="header-section-number">10.5.1</span> Applications</h3>
</div>
<div id="strengths" class="section level3">
<h3><span class="header-section-number">10.5.2</span> Strengths</h3>
</div>
<div id="limitations" class="section level3">
<h3><span class="header-section-number">10.5.3</span> Limitations</h3>
<!--
Not appropriate for phylogenetic analysis!!
-->
</div>
</div>
<div id="exercises" class="section level2">
<h2><span class="header-section-number">10.6</span> Exercises</h2>
<!--
1. Toy clusters
2. mouse mammary time-course (kmeans and dbscan)
3. dimensionality reduction before clustering (helpful for visualization if using a partitioning method) - possibly use parasite data?
4. exercise involving heatmap

-->
<p>Exercise solutions: <a href="solutions-to-exercises.html#solutions-clustering">B.8</a></p>

</div>
</div>



<h3>References</h3>
<div id="refs" class="references">
<div id="ref-Gionis2007">
<p>Gionis, A., H. Mannila, and P. Tsaparas. 2007. “Clustering aggregation.” <em>ACM Transactions on Knowledge Discovery from Data (TKDD)</em> 1 (1): 1–30.</p>
</div>
<div id="ref-Veenman2002">
<p>Veenman, C.J., M.J.T. Reinders, and E. Backer. 2002. “A maximum variance cluster algorithm.” <em>IEEE Trans. Pattern Analysis and Machine Intelligence</em> 24 (9): 1273–80.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="dimensionality-reduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="resources.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/bioinformatics-training/intro-machine-learning/edit/master/09-clustering.Rmd",
"text": "Edit"
},
"download": ["intro-machine-learning.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
