<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>An Introduction to Machine Learning</title>
  <meta name="description" content="Course materials for An Introduction to Machine Learning">
  <meta name="generator" content="bookdown 0.5 and GitBook 2.6.7">

  <meta property="og:title" content="An Introduction to Machine Learning" />
  <meta property="og:type" content="book" />
  
  <meta property="og:image" content="figures/cover_image.png" />
  <meta property="og:description" content="Course materials for An Introduction to Machine Learning" />
  <meta name="github-repo" content="bioinformatics-training/intro-machine-learning" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="An Introduction to Machine Learning" />
  
  <meta name="twitter:description" content="Course materials for An Introduction to Machine Learning" />
  <meta name="twitter:image" content="figures/cover_image.png" />

<meta name="author" content="Sudhakaran Prabakaran, Matt Wayland and Chris Penfold">


<meta name="date" content="2017-09-04">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="dimensionality-reduction.html">
<link rel="next" href="resources.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">An Introduction to Machine Learning</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About the course</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#overview"><i class="fa fa-check"></i><b>1.1</b> Overview</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#registration"><i class="fa fa-check"></i><b>1.2</b> Registration</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.3</b> Prerequisites</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#github"><i class="fa fa-check"></i><b>1.4</b> Github</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.5</b> License</a></li>
<li class="chapter" data-level="1.6" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.6</b> Contact</a></li>
<li class="chapter" data-level="1.7" data-path="index.html"><a href="index.html#colophon"><i class="fa fa-check"></i><b>1.7</b> Colophon</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a></li>
<li class="chapter" data-level="3" data-path="linear-models.html"><a href="linear-models.html"><i class="fa fa-check"></i><b>3</b> Linear models and matrix algebra</a><ul>
<li class="chapter" data-level="3.1" data-path="linear-models.html"><a href="linear-models.html#exercises"><i class="fa fa-check"></i><b>3.1</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="logistic-regression.html"><a href="logistic-regression.html"><i class="fa fa-check"></i><b>4</b> Linear and non linear logistic regression</a><ul>
<li class="chapter" data-level="4.1" data-path="logistic-regression.html"><a href="logistic-regression.html#exercises-1"><i class="fa fa-check"></i><b>4.1</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html"><i class="fa fa-check"></i><b>5</b> Nearest neighbours</a><ul>
<li class="chapter" data-level="5.1" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#example-one"><i class="fa fa-check"></i><b>5.1</b> Example one</a></li>
<li class="chapter" data-level="5.2" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#example-two"><i class="fa fa-check"></i><b>5.2</b> Example two</a></li>
<li class="chapter" data-level="5.3" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#exercises-2"><i class="fa fa-check"></i><b>5.3</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="decision-trees.html"><a href="decision-trees.html"><i class="fa fa-check"></i><b>6</b> Decision trees and random forests</a><ul>
<li class="chapter" data-level="6.1" data-path="decision-trees.html"><a href="decision-trees.html#exercises-3"><i class="fa fa-check"></i><b>6.1</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="svm.html"><a href="svm.html"><i class="fa fa-check"></i><b>7</b> Support vector machines</a><ul>
<li class="chapter" data-level="7.1" data-path="svm.html"><a href="svm.html#exercises-4"><i class="fa fa-check"></i><b>7.1</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="ann.html"><a href="ann.html"><i class="fa fa-check"></i><b>8</b> Artificial neural networks</a><ul>
<li class="chapter" data-level="8.1" data-path="ann.html"><a href="ann.html#exercises-5"><i class="fa fa-check"></i><b>8.1</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html"><i class="fa fa-check"></i><b>9</b> Dimensionality reduction</a><ul>
<li class="chapter" data-level="9.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#linear-dimensionality-reduction"><i class="fa fa-check"></i><b>9.1</b> Linear Dimensionality Reduction</a><ul>
<li class="chapter" data-level="9.1.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#principle-component-analysis"><i class="fa fa-check"></i><b>9.1.1</b> Principle Component Analysis</a></li>
<li class="chapter" data-level="9.1.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#horeshoe-effect"><i class="fa fa-check"></i><b>9.1.2</b> Horeshoe effect</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#nonlinear-dimensionality-reduction"><i class="fa fa-check"></i><b>9.2</b> Nonlinear Dimensionality Reduction</a><ul>
<li class="chapter" data-level="9.2.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#t-sne"><i class="fa fa-check"></i><b>9.2.1</b> t-SNE</a></li>
<li class="chapter" data-level="9.2.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#gaussian-process-latent-variable-models"><i class="fa fa-check"></i><b>9.2.2</b> Gaussian Process Latent Variable Models</a></li>
<li class="chapter" data-level="9.2.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#gplvms-with-informative-priors"><i class="fa fa-check"></i><b>9.2.3</b> GPLVMs with informative priors</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#exercises-6"><i class="fa fa-check"></i><b>9.3</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>10</b> Clustering</a><ul>
<li class="chapter" data-level="10.1" data-path="clustering.html"><a href="clustering.html#introduction"><i class="fa fa-check"></i><b>10.1</b> Introduction</a></li>
<li class="chapter" data-level="10.2" data-path="clustering.html"><a href="clustering.html#distance-metrics"><i class="fa fa-check"></i><b>10.2</b> Distance metrics</a><ul>
<li class="chapter" data-level="10.2.1" data-path="clustering.html"><a href="clustering.html#image-segmentation"><i class="fa fa-check"></i><b>10.2.1</b> Image segmentation</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="clustering.html"><a href="clustering.html#hierarchic-agglomerative"><i class="fa fa-check"></i><b>10.3</b> Hierarchic agglomerative</a><ul>
<li class="chapter" data-level="10.3.1" data-path="clustering.html"><a href="clustering.html#linkage-algorithms"><i class="fa fa-check"></i><b>10.3.1</b> Linkage algorithms</a></li>
<li class="chapter" data-level="10.3.2" data-path="clustering.html"><a href="clustering.html#example-clustering-synthetic-data-sets"><i class="fa fa-check"></i><b>10.3.2</b> Example: clustering synthetic data sets</a></li>
<li class="chapter" data-level="10.3.3" data-path="clustering.html"><a href="clustering.html#example-gene-expression-profiling-of-human-tissues"><i class="fa fa-check"></i><b>10.3.3</b> Example: gene expression profiling of human tissues</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="clustering.html"><a href="clustering.html#k-means"><i class="fa fa-check"></i><b>10.4</b> K-means</a><ul>
<li class="chapter" data-level="10.4.1" data-path="clustering.html"><a href="clustering.html#algorithm"><i class="fa fa-check"></i><b>10.4.1</b> Algorithm</a></li>
<li class="chapter" data-level="10.4.2" data-path="clustering.html"><a href="clustering.html#choosing-initial-cluster-centres"><i class="fa fa-check"></i><b>10.4.2</b> Choosing initial cluster centres</a></li>
<li class="chapter" data-level="10.4.3" data-path="clustering.html"><a href="clustering.html#choosing-k"><i class="fa fa-check"></i><b>10.4.3</b> Choosing k</a></li>
<li class="chapter" data-level="10.4.4" data-path="clustering.html"><a href="clustering.html#example-clustering-synthetic-data-sets-1"><i class="fa fa-check"></i><b>10.4.4</b> Example: clustering synthetic data sets</a></li>
<li class="chapter" data-level="10.4.5" data-path="clustering.html"><a href="clustering.html#example-gene-expression-profiling-of-human-tissues-1"><i class="fa fa-check"></i><b>10.4.5</b> Example: gene expression profiling of human tissues</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="clustering.html"><a href="clustering.html#dbscan"><i class="fa fa-check"></i><b>10.5</b> DBSCAN</a><ul>
<li class="chapter" data-level="10.5.1" data-path="clustering.html"><a href="clustering.html#algorithm-1"><i class="fa fa-check"></i><b>10.5.1</b> Algorithm</a></li>
<li class="chapter" data-level="10.5.2" data-path="clustering.html"><a href="clustering.html#choosing-parameters"><i class="fa fa-check"></i><b>10.5.2</b> Choosing parameters</a></li>
<li class="chapter" data-level="10.5.3" data-path="clustering.html"><a href="clustering.html#example-clustering-synthetic-data-sets-2"><i class="fa fa-check"></i><b>10.5.3</b> Example: clustering synthetic data sets</a></li>
<li class="chapter" data-level="10.5.4" data-path="clustering.html"><a href="clustering.html#gene-expression"><i class="fa fa-check"></i><b>10.5.4</b> Gene expression</a></li>
</ul></li>
<li class="chapter" data-level="10.6" data-path="clustering.html"><a href="clustering.html#summary"><i class="fa fa-check"></i><b>10.6</b> Summary</a><ul>
<li class="chapter" data-level="10.6.1" data-path="clustering.html"><a href="clustering.html#applications"><i class="fa fa-check"></i><b>10.6.1</b> Applications</a></li>
<li class="chapter" data-level="10.6.2" data-path="clustering.html"><a href="clustering.html#strengths"><i class="fa fa-check"></i><b>10.6.2</b> Strengths</a></li>
<li class="chapter" data-level="10.6.3" data-path="clustering.html"><a href="clustering.html#limitations"><i class="fa fa-check"></i><b>10.6.3</b> Limitations</a></li>
</ul></li>
<li class="chapter" data-level="10.7" data-path="clustering.html"><a href="clustering.html#exercises-7"><i class="fa fa-check"></i><b>10.7</b> Exercises</a></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>A</b> Resources</a><ul>
<li class="chapter" data-level="A.1" data-path="resources.html"><a href="resources.html#python"><i class="fa fa-check"></i><b>A.1</b> Python</a></li>
<li class="chapter" data-level="A.2" data-path="resources.html"><a href="resources.html#machine-learning-data-set-repository"><i class="fa fa-check"></i><b>A.2</b> Machine learning data set repository</a></li>
</ul></li>
<li class="chapter" data-level="B" data-path="solutions-linear-models.html"><a href="solutions-linear-models.html"><i class="fa fa-check"></i><b>B</b> Solutions ch. 3 - Linear models and matrix algebra</a><ul>
<li class="chapter" data-level="B.1" data-path="solutions-linear-models.html"><a href="solutions-linear-models.html#exercise-1"><i class="fa fa-check"></i><b>B.1</b> Exercise 1</a></li>
<li class="chapter" data-level="B.2" data-path="solutions-linear-models.html"><a href="solutions-linear-models.html#exercise-2"><i class="fa fa-check"></i><b>B.2</b> Exercise 2</a></li>
</ul></li>
<li class="chapter" data-level="C" data-path="solutions-logistic-regression.html"><a href="solutions-logistic-regression.html"><i class="fa fa-check"></i><b>C</b> Solutions ch. 4 - Linear and non-linear logistic regression</a><ul>
<li class="chapter" data-level="C.1" data-path="solutions-logistic-regression.html"><a href="solutions-logistic-regression.html#exercise-1-1"><i class="fa fa-check"></i><b>C.1</b> Exercise 1</a></li>
<li class="chapter" data-level="C.2" data-path="solutions-logistic-regression.html"><a href="solutions-logistic-regression.html#exercise-2-1"><i class="fa fa-check"></i><b>C.2</b> Exercise 2</a></li>
</ul></li>
<li class="chapter" data-level="D" data-path="solutions-nearest-neighbours.html"><a href="solutions-nearest-neighbours.html"><i class="fa fa-check"></i><b>D</b> Solutions ch. 5 - Nearest neighbours</a><ul>
<li class="chapter" data-level="D.1" data-path="solutions-nearest-neighbours.html"><a href="solutions-nearest-neighbours.html#exercise-1-2"><i class="fa fa-check"></i><b>D.1</b> Exercise 1</a></li>
<li class="chapter" data-level="D.2" data-path="solutions-nearest-neighbours.html"><a href="solutions-nearest-neighbours.html#exercise-2-2"><i class="fa fa-check"></i><b>D.2</b> Exercise 2</a></li>
</ul></li>
<li class="chapter" data-level="E" data-path="solutions-decision-trees.html"><a href="solutions-decision-trees.html"><i class="fa fa-check"></i><b>E</b> Solutions ch. 6 - Decision trees and random forests</a><ul>
<li class="chapter" data-level="E.1" data-path="solutions-decision-trees.html"><a href="solutions-decision-trees.html#exercise-1-3"><i class="fa fa-check"></i><b>E.1</b> Exercise 1</a></li>
<li class="chapter" data-level="E.2" data-path="solutions-decision-trees.html"><a href="solutions-decision-trees.html#exercise-2-3"><i class="fa fa-check"></i><b>E.2</b> Exercise 2</a></li>
</ul></li>
<li class="chapter" data-level="F" data-path="solutions-svm.html"><a href="solutions-svm.html"><i class="fa fa-check"></i><b>F</b> Solutions ch. 7 - Support vector machines</a><ul>
<li class="chapter" data-level="F.1" data-path="solutions-svm.html"><a href="solutions-svm.html#exercise-1-4"><i class="fa fa-check"></i><b>F.1</b> Exercise 1</a></li>
<li class="chapter" data-level="F.2" data-path="solutions-svm.html"><a href="solutions-svm.html#exercise-2-4"><i class="fa fa-check"></i><b>F.2</b> Exercise 2</a></li>
</ul></li>
<li class="chapter" data-level="G" data-path="solutions-ann.html"><a href="solutions-ann.html"><i class="fa fa-check"></i><b>G</b> Solutions ch. 8 - Artificial neural networks</a><ul>
<li class="chapter" data-level="G.1" data-path="solutions-ann.html"><a href="solutions-ann.html#exercise-1-5"><i class="fa fa-check"></i><b>G.1</b> Exercise 1</a></li>
<li class="chapter" data-level="G.2" data-path="solutions-ann.html"><a href="solutions-ann.html#exercise-2-5"><i class="fa fa-check"></i><b>G.2</b> Exercise 2</a></li>
</ul></li>
<li class="chapter" data-level="H" data-path="solutions-dimensionality-reduction.html"><a href="solutions-dimensionality-reduction.html"><i class="fa fa-check"></i><b>H</b> Solutions ch. 9 - Dimensionality reduction</a><ul>
<li class="chapter" data-level="H.1" data-path="solutions-dimensionality-reduction.html"><a href="solutions-dimensionality-reduction.html#exercise-1-6"><i class="fa fa-check"></i><b>H.1</b> Exercise 1</a></li>
<li class="chapter" data-level="H.2" data-path="solutions-dimensionality-reduction.html"><a href="solutions-dimensionality-reduction.html#exercise-2-6"><i class="fa fa-check"></i><b>H.2</b> Exercise 2</a></li>
</ul></li>
<li class="chapter" data-level="I" data-path="solutions-clustering.html"><a href="solutions-clustering.html"><i class="fa fa-check"></i><b>I</b> Solutions ch. 10 - Clustering</a><ul>
<li class="chapter" data-level="I.1" data-path="solutions-clustering.html"><a href="solutions-clustering.html#exercise-1-7"><i class="fa fa-check"></i><b>I.1</b> Exercise 1</a></li>
<li class="chapter" data-level="I.2" data-path="solutions-clustering.html"><a href="solutions-clustering.html#exercise-2-7"><i class="fa fa-check"></i><b>I.2</b> Exercise 2</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">An Introduction to Machine Learning</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="clustering" class="section level1">
<h1><span class="header-section-number">10</span> Clustering</h1>
<!-- Matt -->
<!-- http://www.madgroup.path.cam.ac.uk/microarraysummary.shtml -->
<div id="introduction" class="section level2">
<h2><span class="header-section-number">10.1</span> Introduction</h2>
<p>Hierarchic (produce dendrogram) vs partitioning methods</p>
<ul>
<li>Hierarchic agglomerative</li>
<li>k-means</li>
<li>DBSCAN</li>
</ul>
<div class="figure" style="text-align: center"><span id="fig:clusterTypes"></span>
<img src="09-clustering_files/figure-html/clusterTypes-1.png" alt="Example clusters. **A**, *blobs*; **B**, *aggregation* [@Gionis2007]; **C**, *noisy moons*; **D**, *noisy circles*; **E**, *anisotropic distributions*; **F**, *no structure*." width="80%" />
<p class="caption">
Figure 10.1: Example clusters. <strong>A</strong>, <em>blobs</em>; <strong>B</strong>, <em>aggregation</em> <span class="citation">(Gionis, Mannila, and Tsaparas <a href="#ref-Gionis2007">2007</a>)</span>; <strong>C</strong>, <em>noisy moons</em>; <strong>D</strong>, <em>noisy circles</em>; <strong>E</strong>, <em>anisotropic distributions</em>; <strong>F</strong>, <em>no structure</em>.
</p>
</div>
</div>
<div id="distance-metrics" class="section level2">
<h2><span class="header-section-number">10.2</span> Distance metrics</h2>
<p>dist function cor as.dist(1-cor(x))</p>
<strong>Minkowski distance:</strong>
<span class="math display" id="eq:minkowski">\[\begin{equation}
  distance\left(x,y,p\right)=\left(\sum_{i=1}^{n} abs(x_i-y_i)^p\right)^{1/p}
  \tag{10.1}
\end{equation}\]</span>
<p>Graphical explanation of euclidean, manhattan and max (Chebyshev?)</p>
<div id="image-segmentation" class="section level3">
<h3><span class="header-section-number">10.2.1</span> Image segmentation</h3>
</div>
</div>
<div id="hierarchic-agglomerative" class="section level2">
<h2><span class="header-section-number">10.3</span> Hierarchic agglomerative</h2>
<table>
<caption><span id="tab:distance-matrix">Table 10.1: </span>Example distance matrix</caption>
<thead>
<tr class="header">
<th></th>
<th align="left">A</th>
<th align="left">B</th>
<th align="left">C</th>
<th align="left">D</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>B</td>
<td align="left">2</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="even">
<td>C</td>
<td align="left">6</td>
<td align="left">5</td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="odd">
<td>D</td>
<td align="left">10</td>
<td align="left">10</td>
<td align="left">5</td>
<td align="left"></td>
</tr>
<tr class="even">
<td>E</td>
<td align="left">9</td>
<td align="left">8</td>
<td align="left">3</td>
<td align="left">4</td>
</tr>
</tbody>
</table>
<div id="linkage-algorithms" class="section level3">
<h3><span class="header-section-number">10.3.1</span> Linkage algorithms</h3>
<p>Make one section panel of three dendrograms one table</p>
<p>Single linkage - nearest neighbours linkage Complete linkage - furthest neighbours linkage Average linkage - UPGMA (Unweighted Pair Group Method with Arithmetic Mean)</p>
<!--
Explain anatomy of the dendrogram - branches, nodes and leaves.
-->
<table>
<caption><span id="tab:distance-merge">Table 10.2: </span>Merge distances for objects in the example distance matrix using three different linkage methods.</caption>
<thead>
<tr class="header">
<th align="left">Groups</th>
<th align="left">Single</th>
<th align="left">Complete</th>
<th align="left">Average</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">A,B,C,D,E</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
</tr>
<tr class="even">
<td align="left">(A,B),C,D,E</td>
<td align="left">2</td>
<td align="left">2</td>
<td align="left">2</td>
</tr>
<tr class="odd">
<td align="left">(A,B),(C,E),D</td>
<td align="left">3</td>
<td align="left">3</td>
<td align="left">3</td>
</tr>
<tr class="even">
<td align="left">(A,B)(C,D,E)</td>
<td align="left">4</td>
<td align="left">5</td>
<td align="left">4.5</td>
</tr>
<tr class="odd">
<td align="left">(A,B,C,D,E)</td>
<td align="left">5</td>
<td align="left">10</td>
<td align="left">8</td>
</tr>
</tbody>
</table>
<div class="figure" style="text-align: center"><span id="fig:linkageComparison"></span>
<img src="09-clustering_files/figure-html/linkageComparison-1.png" alt="Dendrograms for the example distance matrix using three different linkage methods. " width="100%" /><img src="09-clustering_files/figure-html/linkageComparison-2.png" alt="Dendrograms for the example distance matrix using three different linkage methods. " width="100%" /><img src="09-clustering_files/figure-html/linkageComparison-3.png" alt="Dendrograms for the example distance matrix using three different linkage methods. " width="100%" />
<p class="caption">
Figure 10.2: Dendrograms for the example distance matrix using three different linkage methods.
</p>
</div>
</div>
<div id="example-clustering-synthetic-data-sets" class="section level3">
<h3><span class="header-section-number">10.3.2</span> Example: clustering synthetic data sets</h3>
<div id="step-by-step-instructions" class="section level4">
<h4><span class="header-section-number">10.3.2.1</span> Step-by-step instructions</h4>
<ol style="list-style-type: decimal">
<li>Load required packages.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RColorBrewer)
<span class="kw">library</span>(dendextend)</code></pre></div>
<pre><code>## 
## ---------------------
## Welcome to dendextend version 1.5.2
## Type citation(&#39;dendextend&#39;) for how to cite the package.
## 
## Type browseVignettes(package = &#39;dendextend&#39;) for the package vignette.
## The github page is: https://github.com/talgalili/dendextend/
## 
## Suggestions and bug-reports can be submitted at: https://github.com/talgalili/dendextend/issues
## Or contact: &lt;tal.galili@gmail.com&gt;
## 
##  To suppress this message use:  suppressPackageStartupMessages(library(dendextend))
## ---------------------</code></pre>
<pre><code>## 
## Attaching package: &#39;dendextend&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:ggdendro&#39;:
## 
##     theme_dendro</code></pre>
<pre><code>## The following object is masked from &#39;package:stats&#39;:
## 
##     cutree</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
<span class="kw">library</span>(GGally)</code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Retrieve a palette of eight colours.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cluster_colours &lt;-<span class="st"> </span><span class="kw">brewer.pal</span>(<span class="dv">8</span>,<span class="st">&quot;Dark2&quot;</span>)</code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Read in data for <strong>blobs</strong> example.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">blobs &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;data/example_clusters/blobs.csv&quot;</span>, <span class="dt">header=</span>F)</code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li>Create distance matrix using Euclidean distance metric.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">d &lt;-<span class="st"> </span><span class="kw">dist</span>(blobs[,<span class="dv">1</span>:<span class="dv">2</span>])</code></pre></div>
<ol start="5" style="list-style-type: decimal">
<li>Perform hierarchical clustering using the <strong>average</strong> agglomeration method and convert the result to an object of class <strong>dendrogram</strong>. A <strong>dendrogram</strong> object can be edited using the advanced features of the <strong>dendextend</strong> package.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dend &lt;-<span class="st"> </span><span class="kw">as.dendrogram</span>(<span class="kw">hclust</span>(d, <span class="dt">method=</span><span class="st">&quot;average&quot;</span>))</code></pre></div>
<ol start="6" style="list-style-type: decimal">
<li>Cut the tree into three clusters</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clusters &lt;-<span class="st"> </span><span class="kw">cutree</span>(dend,<span class="dv">3</span>,<span class="dt">order_clusters_as_data=</span>F)</code></pre></div>
<ol start="7" style="list-style-type: decimal">
<li>The vector <strong>clusters</strong> contains the cluster membership (in this case <em>1</em>, <em>2</em> or <em>3</em>) of each observation (data point) in the order they appear on the dendrogram. We can use this vector to colour the branches of the dendrogram by cluster.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dend &lt;-<span class="st"> </span><span class="kw">color_branches</span>(dend, <span class="dt">clusters=</span>clusters, <span class="dt">col=</span>cluster_colours[<span class="dv">1</span>:<span class="dv">3</span>])</code></pre></div>
<ol start="8" style="list-style-type: decimal">
<li>We can use the <strong>labels</strong> function to annotate the leaves of the dendrogram. However, it is not possible to create legible labels for the 1,500 leaves in our example dendrogram, so we will set the label for each leaf to an empty string.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">labels</span>(dend) &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="st">&quot;&quot;</span>, <span class="kw">length</span>(blobs[,<span class="dv">1</span>]))</code></pre></div>
<ol start="9" style="list-style-type: decimal">
<li>If we want to plot the dendrogram using <strong>ggplot</strong>, we must convert it to an object of class <strong>ggdend</strong>.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ggd &lt;-<span class="st"> </span><span class="kw">as.ggdend</span>(dend)</code></pre></div>
<ol start="10" style="list-style-type: decimal">
<li>The <strong>nodes</strong> attribute of <strong>ggd</strong> is a data.frame of parameters related to the plotting of dendogram nodes. The <strong>nodes</strong> data.frame contains some NAs which will generate warning messages when <strong>ggd</strong> is processed by <strong>ggplot</strong>. Since we are not interested in annotating dendrogram nodes, the easiest option here is to delete all of the rows of <strong>nodes</strong>.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ggd$nodes &lt;-<span class="st"> </span>ggd$nodes[!(<span class="dv">1</span>:<span class="kw">length</span>(ggd$nodes[,<span class="dv">1</span>])),]</code></pre></div>
<ol start="11" style="list-style-type: decimal">
<li>We can use the cluster membership of each observation contained in the vector <strong>clusters</strong> to assign colours to the data points of a scatterplot. However, first we need to reorder the vector so that the cluster memberships are in the same order that the observations appear in the data.frame of observations. Fortunately the names of the elements of the vector are the indices of the observations in the data.frame and so reordering can be accomplished in one line.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clusters &lt;-<span class="st"> </span>clusters[<span class="kw">order</span>(<span class="kw">as.numeric</span>(<span class="kw">names</span>(clusters)))]</code></pre></div>
<ol start="12" style="list-style-type: decimal">
<li>We are now ready to plot a dendrogram and scatterplot. We will use the <strong>ggmatrix</strong> function from the <strong>GGally</strong> package to place the plots side-by-side.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">plotList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="kw">ggplot</span>(ggd),
                 <span class="kw">ggplot</span>(blobs, <span class="kw">aes</span>(V1,V2)) +<span class="st"> </span>
<span class="st">                   </span><span class="kw">geom_point</span>(<span class="dt">col=</span>cluster_colours[clusters], <span class="dt">size=</span><span class="fl">0.2</span>)
                 )

pm &lt;-<span class="st"> </span><span class="kw">ggmatrix</span>(
  plotList, <span class="dt">nrow=</span><span class="dv">1</span>, <span class="dt">ncol=</span><span class="dv">2</span>, <span class="dt">showXAxisPlotLabels =</span> F, <span class="dt">showYAxisPlotLabels =</span> F, 
  <span class="dt">xAxisLabels=</span><span class="kw">c</span>(<span class="st">&quot;dendrogram&quot;</span>, <span class="st">&quot;scatter plot&quot;</span>)
) +<span class="st"> </span><span class="kw">theme_bw</span>()

pm</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:hclustBlobs"></span>
<img src="09-clustering_files/figure-html/hclustBlobs-1.png" alt="Hierarchical clustering of the blobs data set." width="80%" />
<p class="caption">
Figure 10.3: Hierarchical clustering of the blobs data set.
</p>
</div>
</div>
<div id="clustering-of-other-synthetic-data-sets" class="section level4">
<h4><span class="header-section-number">10.3.2.2</span> Clustering of other synthetic data sets</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">aggregation &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/example_clusters/aggregation.txt&quot;</span>)
noisy_moons &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;data/example_clusters/noisy_moons.csv&quot;</span>, <span class="dt">header=</span>F)
noisy_circles &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;data/example_clusters/noisy_circles.csv&quot;</span>, <span class="dt">header=</span>F)
aniso &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;data/example_clusters/aniso.csv&quot;</span>, <span class="dt">header=</span>F)
no_structure &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;data/example_clusters/no_structure.csv&quot;</span>, <span class="dt">header=</span>F)

hclust_plots &lt;-<span class="st"> </span>function(data_set, n){
  d &lt;-<span class="st"> </span><span class="kw">dist</span>(data_set[,<span class="dv">1</span>:<span class="dv">2</span>])
  dend &lt;-<span class="st"> </span><span class="kw">as.dendrogram</span>(<span class="kw">hclust</span>(d, <span class="dt">method=</span><span class="st">&quot;average&quot;</span>))
  clusters &lt;-<span class="st"> </span><span class="kw">cutree</span>(dend,n,<span class="dt">order_clusters_as_data=</span>F)
  dend &lt;-<span class="st"> </span><span class="kw">color_branches</span>(dend, <span class="dt">clusters=</span>clusters, <span class="dt">col=</span>cluster_colours[<span class="dv">1</span>:n])
  clusters &lt;-<span class="st"> </span>clusters[<span class="kw">order</span>(<span class="kw">as.numeric</span>(<span class="kw">names</span>(clusters)))]
  <span class="kw">labels</span>(dend) &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="st">&quot;&quot;</span>, <span class="kw">length</span>(data_set[,<span class="dv">1</span>]))
  ggd &lt;-<span class="st"> </span><span class="kw">as.ggdend</span>(dend)
  ggd$nodes &lt;-<span class="st"> </span>ggd$nodes[!(<span class="dv">1</span>:<span class="kw">length</span>(ggd$nodes[,<span class="dv">1</span>])),]
  plotPair &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="kw">ggplot</span>(ggd),
    <span class="kw">ggplot</span>(data_set, <span class="kw">aes</span>(V1,V2)) +<span class="st"> </span>
<span class="st">      </span><span class="kw">geom_point</span>(<span class="dt">col=</span>cluster_colours[clusters], <span class="dt">size=</span><span class="fl">0.2</span>))
  <span class="kw">return</span>(plotPair)
}

plotList &lt;-<span class="st"> </span><span class="kw">c</span>(
  <span class="kw">hclust_plots</span>(aggregation, <span class="dv">7</span>),
  <span class="kw">hclust_plots</span>(noisy_moons, <span class="dv">2</span>),
  <span class="kw">hclust_plots</span>(noisy_circles, <span class="dv">2</span>),
  <span class="kw">hclust_plots</span>(aniso, <span class="dv">3</span>),
  <span class="kw">hclust_plots</span>(no_structure, <span class="dv">3</span>)
)

pm &lt;-<span class="st"> </span><span class="kw">ggmatrix</span>(
  plotList, <span class="dt">nrow=</span><span class="dv">5</span>, <span class="dt">ncol=</span><span class="dv">2</span>, <span class="dt">showXAxisPlotLabels =</span> F, <span class="dt">showYAxisPlotLabels =</span> F,
  <span class="dt">xAxisLabels=</span><span class="kw">c</span>(<span class="st">&quot;dendrogram&quot;</span>, <span class="st">&quot;scatter plot&quot;</span>), 
  <span class="dt">yAxisLabels=</span><span class="kw">c</span>(<span class="st">&quot;aggregation&quot;</span>, <span class="st">&quot;noisy moons&quot;</span>, <span class="st">&quot;noisy circles&quot;</span>, <span class="st">&quot;anisotropic&quot;</span>, <span class="st">&quot;no structure&quot;</span>)
) +<span class="st"> </span><span class="kw">theme_bw</span>()

pm</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:hclustToyData"></span>
<img src="09-clustering_files/figure-html/hclustToyData-1.png" alt="Hierarchical clustering of synthetic data-sets. " width="75%" />
<p class="caption">
Figure 10.4: Hierarchical clustering of synthetic data-sets.
</p>
</div>
</div>
</div>
<div id="example-gene-expression-profiling-of-human-tissues" class="section level3">
<h3><span class="header-section-number">10.3.3</span> Example: gene expression profiling of human tissues</h3>
<div id="basics" class="section level4">
<h4><span class="header-section-number">10.3.3.1</span> Basics</h4>
<p>Load required libraries</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RColorBrewer)
<span class="kw">library</span>(dendextend)</code></pre></div>
<p>Load data</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&quot;data/tissues_gene_expression/tissuesGeneExpression.rda&quot;</span>)</code></pre></div>
<p>Inspect data</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(tissue)</code></pre></div>
<pre><code>## tissue
##  cerebellum       colon endometrium hippocampus      kidney       liver 
##          38          34          15          31          39          26 
##    placenta 
##           6</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(e)</code></pre></div>
<pre><code>## [1] 22215   189</code></pre>
<p>Compute distance between each sample</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">d &lt;-<span class="st"> </span><span class="kw">dist</span>(<span class="kw">t</span>(e))</code></pre></div>
<p>perform hierarchical clustering</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc &lt;-<span class="st"> </span><span class="kw">hclust</span>(d, <span class="dt">method=</span><span class="st">&quot;average&quot;</span>)
<span class="kw">plot</span>(hc, <span class="dt">labels=</span>tissue, <span class="dt">cex=</span><span class="fl">0.5</span>, <span class="dt">hang=</span>-<span class="dv">1</span>, <span class="dt">xlab=</span><span class="st">&quot;&quot;</span>, <span class="dt">sub=</span><span class="st">&quot;&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:tissueDendrogram"></span>
<img src="09-clustering_files/figure-html/tissueDendrogram-1.png" alt="Clustering of tissue samples based on gene expression profiles. " width="100%" />
<p class="caption">
Figure 10.5: Clustering of tissue samples based on gene expression profiles.
</p>
</div>
</div>
<div id="colour-labels" class="section level4">
<h4><span class="header-section-number">10.3.3.2</span> Colour labels</h4>
<p>use dendextend library to plot dendrogram with colour labels</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tissue_type &lt;-<span class="st"> </span><span class="kw">unique</span>(tissue)
dend &lt;-<span class="st"> </span><span class="kw">as.dendrogram</span>(hc)
dend_colours &lt;-<span class="st"> </span><span class="kw">brewer.pal</span>(<span class="kw">length</span>(<span class="kw">unique</span>(tissue)),<span class="st">&quot;Dark2&quot;</span>)
<span class="kw">names</span>(dend_colours) &lt;-<span class="st"> </span>tissue_type
<span class="kw">labels</span>(dend) &lt;-<span class="st"> </span>tissue[<span class="kw">order.dendrogram</span>(dend)]
<span class="kw">labels_colors</span>(dend) &lt;-<span class="st"> </span>dend_colours[tissue][<span class="kw">order.dendrogram</span>(dend)]
<span class="kw">labels_cex</span>(dend) =<span class="st"> </span><span class="fl">0.5</span>
<span class="kw">plot</span>(dend, <span class="dt">horiz=</span>T)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:tissueDendrogramColour"></span>
<img src="09-clustering_files/figure-html/tissueDendrogramColour-1.png" alt="Clustering of tissue samples based on gene expression profiles with labels coloured by tissue type. " width="100%" />
<p class="caption">
Figure 10.6: Clustering of tissue samples based on gene expression profiles with labels coloured by tissue type.
</p>
</div>
</div>
<div id="defining-clusters-by-cutting-tree" class="section level4">
<h4><span class="header-section-number">10.3.3.3</span> Defining clusters by cutting tree</h4>
<p>Define clusters by cutting tree at a specific height</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(dend, <span class="dt">horiz=</span>T)
<span class="kw">abline</span>(<span class="dt">v=</span><span class="dv">125</span>, <span class="dt">lwd=</span><span class="dv">2</span>, <span class="dt">lty=</span><span class="dv">2</span>, <span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:tissueDendrogramCutHeight"></span>
<img src="09-clustering_files/figure-html/tissueDendrogramCutHeight-1.png" alt="Clusters found by cutting tree at a height of 125" width="100%" />
<p class="caption">
Figure 10.7: Clusters found by cutting tree at a height of 125
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hclusters &lt;-<span class="st"> </span><span class="kw">cutree</span>(dend, <span class="dt">h=</span><span class="dv">125</span>)
<span class="kw">table</span>(tissue, <span class="dt">cluster=</span>hclusters)</code></pre></div>
<pre><code>##              cluster
## tissue         1  2  3  4  5  6
##   cerebellum   0 36  0  0  2  0
##   colon        0  0 34  0  0  0
##   endometrium 15  0  0  0  0  0
##   hippocampus  0 31  0  0  0  0
##   kidney      37  0  0  0  2  0
##   liver        0  0  0 24  2  0
##   placenta     0  0  0  0  0  6</code></pre>
<p>Select a specific number of clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(dend, <span class="dt">horiz=</span>T)
<span class="kw">abline</span>(<span class="dt">v =</span> <span class="kw">heights_per_k.dendrogram</span>(dend)[<span class="st">&quot;8&quot;</span>], <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">lty =</span> <span class="dv">2</span>, <span class="dt">col =</span> <span class="st">&quot;blue&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:tissueDendrogramEightClusters"></span>
<img src="09-clustering_files/figure-html/tissueDendrogramEightClusters-1.png" alt="Selection of eight clusters from the dendogram" width="100%" />
<p class="caption">
Figure 10.8: Selection of eight clusters from the dendogram
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hclusters &lt;-<span class="st"> </span><span class="kw">cutree</span>(dend, <span class="dt">k=</span><span class="dv">8</span>)
<span class="kw">table</span>(tissue, <span class="dt">cluster=</span>hclusters)</code></pre></div>
<pre><code>##              cluster
## tissue         1  2  3  4  5  6  7  8
##   cerebellum   0 31  0  0  2  0  5  0
##   colon        0  0 34  0  0  0  0  0
##   endometrium  0  0  0  0  0 15  0  0
##   hippocampus  0 31  0  0  0  0  0  0
##   kidney      37  0  0  0  2  0  0  0
##   liver        0  0  0 24  2  0  0  0
##   placenta     0  0  0  0  0  0  0  6</code></pre>
</div>
<div id="heatmap" class="section level4">
<h4><span class="header-section-number">10.3.3.4</span> Heatmap</h4>
<p>Base R provides a <strong>heatmap</strong> function, but we will use the more advanced <strong>heatmap.2</strong> from the <strong>gplots</strong> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(gplots)</code></pre></div>
<pre><code>## 
## Attaching package: &#39;gplots&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:stats&#39;:
## 
##     lowess</code></pre>
<p>Define a colour palette (also known as a lookup table).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">heatmap_colours &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>(<span class="kw">brewer.pal</span>(<span class="dv">9</span>, <span class="st">&quot;PuBuGn&quot;</span>))(<span class="dv">100</span>)</code></pre></div>
<p>Calculate the variance of each gene.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">geneVariance &lt;-<span class="st"> </span><span class="kw">apply</span>(e,<span class="dv">1</span>,var)</code></pre></div>
<p>Find the row numbers of the 40 genes with the highest variance.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">idxTop40 &lt;-<span class="st"> </span><span class="kw">order</span>(-geneVariance)[<span class="dv">1</span>:<span class="dv">40</span>]</code></pre></div>
<p>Define colours for tissues.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tissueColours &lt;-<span class="st"> </span><span class="kw">palette</span>(<span class="kw">brewer.pal</span>(<span class="dv">8</span>, <span class="st">&quot;Dark2&quot;</span>))[<span class="kw">as.numeric</span>(<span class="kw">as.factor</span>(tissue))]</code></pre></div>
<p>Plot heatmap.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">heatmap.2</span>(e[idxTop40,], <span class="dt">labCol=</span>tissue, <span class="dt">trace=</span><span class="st">&quot;none&quot;</span>,
          <span class="dt">ColSideColors=</span>tissueColours, <span class="dt">col=</span>heatmap_colours)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:heatmapTissueExpression"></span>
<img src="09-clustering_files/figure-html/heatmapTissueExpression-1.png" alt="Heatmap of the expression of the 40 genes with the highest variance." width="100%" />
<p class="caption">
Figure 10.9: Heatmap of the expression of the 40 genes with the highest variance.
</p>
</div>
</div>
</div>
</div>
<div id="k-means" class="section level2">
<h2><span class="header-section-number">10.4</span> K-means</h2>
<div id="algorithm" class="section level3">
<h3><span class="header-section-number">10.4.1</span> Algorithm</h3>
<p>Pseudocode</p>
<p>to illustrate range of different types of data that can be clustered - image segmentation</p>
<div class="figure" style="text-align: center"><span id="fig:kmeansIterations"></span>
<img src="09-clustering_files/figure-html/kmeansIterations-1.png" alt="Iterations of the k-means algorithm" width="90%" />
<p class="caption">
Figure 10.10: Iterations of the k-means algorithm
</p>
</div>
<p>The default setting of the <strong>kmeans</strong> function is to perform a maximum of 10 iterations and if the algorithm fails to converge a warning is issued. The maximum number of iterations is set with the argument <strong>iter.max</strong>.</p>
</div>
<div id="choosing-initial-cluster-centres" class="section level3">
<h3><span class="header-section-number">10.4.2</span> Choosing initial cluster centres</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RColorBrewer)
point_shapes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">15</span>,<span class="dv">17</span>,<span class="dv">19</span>)
point_colours &lt;-<span class="st"> </span><span class="kw">brewer.pal</span>(<span class="dv">3</span>,<span class="st">&quot;Dark2&quot;</span>)
point_size =<span class="st"> </span><span class="fl">1.5</span>
center_point_size =<span class="st"> </span><span class="dv">8</span>

blobs &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">read.csv</span>(<span class="st">&quot;data/example_clusters/blobs.csv&quot;</span>, <span class="dt">header=</span>F))

good_centres &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">8</span>,<span class="dv">7</span>,<span class="dv">3</span>,<span class="dv">12</span>,<span class="dv">7</span>), <span class="dt">ncol=</span><span class="dv">2</span>, <span class="dt">byrow=</span>T))
bad_centres &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">13</span>,<span class="dv">13</span>,<span class="dv">8</span>,<span class="dv">12</span>,<span class="dv">2</span>,<span class="dv">2</span>), <span class="dt">ncol=</span><span class="dv">2</span>, <span class="dt">byrow=</span>T))

good_result &lt;-<span class="st"> </span><span class="kw">kmeans</span>(blobs[,<span class="dv">1</span>:<span class="dv">2</span>], <span class="dt">centers=</span>good_centres)
bad_result &lt;-<span class="st"> </span><span class="kw">kmeans</span>(blobs[,<span class="dv">1</span>:<span class="dv">2</span>], <span class="dt">centers=</span>bad_centres)

plotList &lt;-<span class="st"> </span><span class="kw">list</span>(
<span class="kw">ggplot</span>(blobs, <span class="kw">aes</span>(V1,V2)) +<span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">col=</span>point_colours[good_result$cluster], <span class="dt">shape=</span>point_shapes[good_result$cluster], 
             <span class="dt">size=</span>point_size) +<span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">data=</span>good_centres, <span class="kw">aes</span>(V1,V2), <span class="dt">shape=</span><span class="dv">3</span>, <span class="dt">col=</span><span class="st">&quot;black&quot;</span>, <span class="dt">size=</span>center_point_size) +<span class="st"> </span>
<span class="st">  </span><span class="kw">theme_bw</span>(),
<span class="kw">ggplot</span>(blobs, <span class="kw">aes</span>(V1,V2)) +<span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">col=</span>point_colours[bad_result$cluster], <span class="dt">shape=</span>point_shapes[bad_result$cluster], 
             <span class="dt">size=</span>point_size) +<span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">data=</span>bad_centres, <span class="kw">aes</span>(V1,V2), <span class="dt">shape=</span><span class="dv">3</span>, <span class="dt">col=</span><span class="st">&quot;black&quot;</span>, <span class="dt">size=</span>center_point_size) +<span class="st"> </span>
<span class="st">  </span><span class="kw">theme_bw</span>()
)

pm &lt;-<span class="st"> </span><span class="kw">ggmatrix</span>(
  plotList, <span class="dt">nrow=</span><span class="dv">1</span>, <span class="dt">ncol=</span><span class="dv">2</span>, <span class="dt">showXAxisPlotLabels =</span> T, <span class="dt">showYAxisPlotLabels =</span> T, 
  <span class="dt">xAxisLabels=</span><span class="kw">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;B&quot;</span>)
) +<span class="st"> </span><span class="kw">theme_bw</span>()

pm</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:kmeansCentreChoice"></span>
<img src="09-clustering_files/figure-html/kmeansCentreChoice-1.png" alt="Initial centres determine clusters. The starting centres are shown as crosses. **A**, real clusters found; **B**, convergence to a local minimum." width="100%" />
<p class="caption">
Figure 10.11: Initial centres determine clusters. The starting centres are shown as crosses. <strong>A</strong>, real clusters found; <strong>B</strong>, convergence to a local minimum.
</p>
</div>
<p>Convergence to a local minimum can be avoided by starting the algorithm multiple times, with different random centres. The <strong>nstart</strong> argument to the <strong>k-means</strong> function can be used to specify the number of random sets and optimal solution will be selected automatically.</p>
</div>
<div id="choosing-k" class="section level3">
<h3><span class="header-section-number">10.4.3</span> Choosing k</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">point_colours &lt;-<span class="st"> </span><span class="kw">brewer.pal</span>(<span class="dv">9</span>,<span class="st">&quot;Set1&quot;</span>)
k &lt;-<span class="st"> </span><span class="dv">1</span>:<span class="dv">9</span>
res &lt;-<span class="st"> </span><span class="kw">lapply</span>(k, function(i){<span class="kw">kmeans</span>(blobs[,<span class="dv">1</span>:<span class="dv">2</span>], i, <span class="dt">nstart=</span><span class="dv">50</span>)})

plotList &lt;-<span class="st"> </span><span class="kw">lapply</span>(k, function(i){
  <span class="kw">ggplot</span>(blobs, <span class="kw">aes</span>(V1, V2)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">col=</span>point_colours[res[[i]]$cluster], <span class="dt">size=</span><span class="dv">1</span>) +
<span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">data=</span><span class="kw">as.data.frame</span>(res[[i]]$centers), <span class="kw">aes</span>(V1,V2), <span class="dt">shape=</span><span class="dv">3</span>, <span class="dt">col=</span><span class="st">&quot;black&quot;</span>, <span class="dt">size=</span><span class="dv">5</span>) +
<span class="st">    </span><span class="kw">annotate</span>(<span class="st">&quot;text&quot;</span>, <span class="dt">x=</span><span class="dv">2</span>, <span class="dt">y=</span><span class="dv">13</span>, <span class="dt">label=</span><span class="kw">paste</span>(<span class="st">&quot;k=&quot;</span>, i, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>), <span class="dt">size=</span><span class="dv">8</span>, <span class="dt">col=</span><span class="st">&quot;black&quot;</span>) +
<span class="st">    </span><span class="kw">theme_bw</span>()
}
)

pm &lt;-<span class="st"> </span><span class="kw">ggmatrix</span>(
  plotList, <span class="dt">nrow=</span><span class="dv">3</span>, <span class="dt">ncol=</span><span class="dv">3</span>, <span class="dt">showXAxisPlotLabels =</span> T, <span class="dt">showYAxisPlotLabels =</span> T
) +<span class="st"> </span><span class="kw">theme_bw</span>()

pm</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:kmeansRangeK"></span>
<img src="09-clustering_files/figure-html/kmeansRangeK-1.png" alt="K-means clustering of the blobs data set using a range of values of k from 1-9. Cluster centres indicated with a cross." width="100%" />
<p class="caption">
Figure 10.12: K-means clustering of the blobs data set using a range of values of k from 1-9. Cluster centres indicated with a cross.
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tot_withinss &lt;-<span class="st"> </span><span class="kw">sapply</span>(k, function(i){res[[i]]$tot.withinss})
<span class="kw">qplot</span>(k, tot_withinss, <span class="dt">geom=</span><span class="kw">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;line&quot;</span>), 
      <span class="dt">ylab=</span><span class="st">&quot;Total within-cluster sum of squares&quot;</span>) +<span class="st"> </span><span class="kw">theme_bw</span>()</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:choosingK"></span>
<img src="09-clustering_files/figure-html/choosingK-1.png" alt="Variance within the clusters. Total within-cluster sum of squares plotted against k." width="50%" />
<p class="caption">
Figure 10.13: Variance within the clusters. Total within-cluster sum of squares plotted against k.
</p>
</div>
<p><em>N.B.</em> we have set <code>nstart=50</code> to run the algorithm 50 times, starting from different, random sets of centroids.</p>
</div>
<div id="example-clustering-synthetic-data-sets-1" class="section level3">
<h3><span class="header-section-number">10.4.4</span> Example: clustering synthetic data sets</h3>
<p>Let’s see how k-means performs on the other toy data sets. First we will define some variables and functions we will use in the analysis of all data sets.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">k=<span class="dv">1</span>:<span class="dv">9</span>
point_shapes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">15</span>,<span class="dv">17</span>,<span class="dv">19</span>,<span class="dv">5</span>,<span class="dv">6</span>,<span class="dv">0</span>,<span class="dv">1</span>)
point_colours &lt;-<span class="st"> </span><span class="kw">brewer.pal</span>(<span class="dv">7</span>,<span class="st">&quot;Dark2&quot;</span>)
point_size =<span class="st"> </span><span class="fl">1.5</span>
center_point_size =<span class="st"> </span><span class="dv">8</span>

plot_tot_withinss &lt;-<span class="st"> </span>function(kmeans_output){
  tot_withinss &lt;-<span class="st"> </span><span class="kw">sapply</span>(k, function(i){kmeans_output[[i]]$tot.withinss})
  <span class="kw">qplot</span>(k, tot_withinss, <span class="dt">geom=</span><span class="kw">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;line&quot;</span>), 
        <span class="dt">ylab=</span><span class="st">&quot;Total within-cluster sum of squares&quot;</span>) +<span class="st"> </span><span class="kw">theme_bw</span>()
}

plot_clusters &lt;-<span class="st"> </span>function(data_set, kmeans_output, num_clusters){
    <span class="kw">ggplot</span>(data_set, <span class="kw">aes</span>(V1,V2)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">col=</span>point_colours[kmeans_output[[num_clusters]]$cluster],
               <span class="dt">shape=</span>point_shapes[kmeans_output[[num_clusters]]$cluster], 
               <span class="dt">size=</span>point_size) +
<span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">data=</span><span class="kw">as.data.frame</span>(kmeans_output[[num_clusters]]$centers), <span class="kw">aes</span>(V1,V2),
               <span class="dt">shape=</span><span class="dv">3</span>,<span class="dt">col=</span><span class="st">&quot;black&quot;</span>,<span class="dt">size=</span>center_point_size) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme_bw</span>()
}</code></pre></div>
<div id="aggregation" class="section level4">
<h4><span class="header-section-number">10.4.4.1</span> Aggregation</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">aggregation &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">read.table</span>(<span class="st">&quot;data/example_clusters/aggregation.txt&quot;</span>))
res &lt;-<span class="st"> </span><span class="kw">lapply</span>(k, function(i){<span class="kw">kmeans</span>(aggregation[,<span class="dv">1</span>:<span class="dv">2</span>], i, <span class="dt">nstart=</span><span class="dv">50</span>)})
<span class="kw">plot_tot_withinss</span>(res)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:kmeansAggregationElbow"></span>
<img src="09-clustering_files/figure-html/kmeansAggregationElbow-1.png" alt="K-means clustering of the aggregation data set: variance within clusters." width="50%" />
<p class="caption">
Figure 10.14: K-means clustering of the aggregation data set: variance within clusters.
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">plotList &lt;-<span class="st"> </span><span class="kw">list</span>(
  <span class="kw">plot_clusters</span>(aggregation, res, <span class="dv">3</span>),
  <span class="kw">plot_clusters</span>(aggregation, res, <span class="dv">7</span>)
)
pm &lt;-<span class="st"> </span><span class="kw">ggmatrix</span>(
  plotList, <span class="dt">nrow=</span><span class="dv">1</span>, <span class="dt">ncol=</span><span class="dv">2</span>, <span class="dt">showXAxisPlotLabels =</span> T, <span class="dt">showYAxisPlotLabels =</span> T, 
  <span class="dt">xAxisLabels=</span><span class="kw">c</span>(<span class="st">&quot;k=3&quot;</span>, <span class="st">&quot;k=7&quot;</span>)
) +<span class="st"> </span><span class="kw">theme_bw</span>()
pm</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:kmeansAggregationScatter"></span>
<img src="09-clustering_files/figure-html/kmeansAggregationScatter-1.png" alt="K-means clustering of the aggregation data set: scatterplots of clusters for k=3 and k=7. Cluster centres indicated with a cross." width="100%" />
<p class="caption">
Figure 10.15: K-means clustering of the aggregation data set: scatterplots of clusters for k=3 and k=7. Cluster centres indicated with a cross.
</p>
</div>
</div>
<div id="noisy-moons" class="section level4">
<h4><span class="header-section-number">10.4.4.2</span> Noisy moons</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">noisy_moons &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;data/example_clusters/noisy_moons.csv&quot;</span>, <span class="dt">header=</span>F)
res &lt;-<span class="st"> </span><span class="kw">lapply</span>(k, function(i){<span class="kw">kmeans</span>(noisy_moons[,<span class="dv">1</span>:<span class="dv">2</span>], i, <span class="dt">nstart=</span><span class="dv">50</span>)})
<span class="kw">plot_tot_withinss</span>(res)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:kmeansNoisyMoonsElbow"></span>
<img src="09-clustering_files/figure-html/kmeansNoisyMoonsElbow-1.png" alt="K-means clustering of the noisy moons data set: variance within clusters." width="50%" />
<p class="caption">
Figure 10.16: K-means clustering of the noisy moons data set: variance within clusters.
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_clusters</span>(noisy_moons, res, <span class="dv">2</span>)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:kmeansNoisyMoonsScatter"></span>
<img src="09-clustering_files/figure-html/kmeansNoisyMoonsScatter-1.png" alt="K-means clustering of the noisy moons data set: scatterplot of clusters for k=2. Cluster centres indicated with a cross." width="50%" />
<p class="caption">
Figure 10.17: K-means clustering of the noisy moons data set: scatterplot of clusters for k=2. Cluster centres indicated with a cross.
</p>
</div>
</div>
<div id="noisy-circles" class="section level4">
<h4><span class="header-section-number">10.4.4.3</span> Noisy circles</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">noisy_circles &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">read.csv</span>(<span class="st">&quot;data/example_clusters/noisy_circles.csv&quot;</span>, <span class="dt">header=</span>F))
res &lt;-<span class="st"> </span><span class="kw">lapply</span>(k, function(i){<span class="kw">kmeans</span>(noisy_circles[,<span class="dv">1</span>:<span class="dv">2</span>], i, <span class="dt">nstart=</span><span class="dv">50</span>)})
<span class="kw">plot_tot_withinss</span>(res)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:kmeansNoisyCirclesElbow"></span>
<img src="09-clustering_files/figure-html/kmeansNoisyCirclesElbow-1.png" alt="K-means clustering of the noisy circles data set: variance within clusters." width="50%" />
<p class="caption">
Figure 10.18: K-means clustering of the noisy circles data set: variance within clusters.
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">plotList &lt;-<span class="st"> </span><span class="kw">list</span>(
  <span class="kw">plot_clusters</span>(noisy_circles, res, <span class="dv">2</span>),
  <span class="kw">plot_clusters</span>(noisy_circles, res, <span class="dv">3</span>)
)
pm &lt;-<span class="st"> </span><span class="kw">ggmatrix</span>(
  plotList, <span class="dt">nrow=</span><span class="dv">1</span>, <span class="dt">ncol=</span><span class="dv">2</span>, <span class="dt">showXAxisPlotLabels =</span> T, 
  <span class="dt">showYAxisPlotLabels =</span> T, <span class="dt">xAxisLabels=</span><span class="kw">c</span>(<span class="st">&quot;k=2&quot;</span>, <span class="st">&quot;k=3&quot;</span>)
) +<span class="st"> </span><span class="kw">theme_bw</span>()
pm</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:kmeansNoisyCirclesScatter"></span>
<img src="09-clustering_files/figure-html/kmeansNoisyCirclesScatter-1.png" alt="K-means clustering of the noisy circles data set: scatterplots of clusters for k=2 and k=3. Cluster centres indicated with a cross." width="100%" />
<p class="caption">
Figure 10.19: K-means clustering of the noisy circles data set: scatterplots of clusters for k=2 and k=3. Cluster centres indicated with a cross.
</p>
</div>
</div>
<div id="anisotropic-distributions" class="section level4">
<h4><span class="header-section-number">10.4.4.4</span> Anisotropic distributions</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">aniso &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">read.csv</span>(<span class="st">&quot;data/example_clusters/aniso.csv&quot;</span>, <span class="dt">header=</span>F))
res &lt;-<span class="st"> </span><span class="kw">lapply</span>(k, function(i){<span class="kw">kmeans</span>(aniso[,<span class="dv">1</span>:<span class="dv">2</span>], i, <span class="dt">nstart=</span><span class="dv">50</span>)})
<span class="kw">plot_tot_withinss</span>(res)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:kmeansAnisoElbow"></span>
<img src="09-clustering_files/figure-html/kmeansAnisoElbow-1.png" alt="K-means clustering  of the anisotropic distributions data set: variance within clusters." width="50%" />
<p class="caption">
Figure 10.20: K-means clustering of the anisotropic distributions data set: variance within clusters.
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">plotList &lt;-<span class="st"> </span><span class="kw">list</span>(
  <span class="kw">plot_clusters</span>(aniso, res, <span class="dv">2</span>),
  <span class="kw">plot_clusters</span>(aniso, res, <span class="dv">3</span>)
)
pm &lt;-<span class="st"> </span><span class="kw">ggmatrix</span>(
  plotList, <span class="dt">nrow=</span><span class="dv">1</span>, <span class="dt">ncol=</span><span class="dv">2</span>, <span class="dt">showXAxisPlotLabels =</span> T, 
  <span class="dt">showYAxisPlotLabels =</span> T, <span class="dt">xAxisLabels=</span><span class="kw">c</span>(<span class="st">&quot;k=2&quot;</span>, <span class="st">&quot;k=3&quot;</span>)
) +<span class="st"> </span><span class="kw">theme_bw</span>()
pm</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:kmeansAnisoScatter"></span>
<img src="09-clustering_files/figure-html/kmeansAnisoScatter-1.png" alt="K-means clustering of the anisotropic distributions data set: scatterplots of clusters for k=2 and k=3. Cluster centres indicated with a cross." width="100%" />
<p class="caption">
Figure 10.21: K-means clustering of the anisotropic distributions data set: scatterplots of clusters for k=2 and k=3. Cluster centres indicated with a cross.
</p>
</div>
</div>
<div id="no-structure" class="section level4">
<h4><span class="header-section-number">10.4.4.5</span> No structure</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">no_structure &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">read.csv</span>(<span class="st">&quot;data/example_clusters/no_structure.csv&quot;</span>, <span class="dt">header=</span>F))
res &lt;-<span class="st"> </span><span class="kw">lapply</span>(k, function(i){<span class="kw">kmeans</span>(no_structure[,<span class="dv">1</span>:<span class="dv">2</span>], i, <span class="dt">nstart=</span><span class="dv">50</span>)})
<span class="kw">plot_tot_withinss</span>(res)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:noStructureElbow"></span>
<img src="09-clustering_files/figure-html/noStructureElbow-1.png" alt="K-means clustering of the data set with no structure: variance within clusters." width="50%" />
<p class="caption">
Figure 10.22: K-means clustering of the data set with no structure: variance within clusters.
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_clusters</span>(no_structure, res, <span class="dv">4</span>)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:noStructureScatter"></span>
<img src="09-clustering_files/figure-html/noStructureScatter-1.png" alt="K-means clustering of the data set with no structure: scatterplot of clusters for k=4. Cluster centres indicated with a cross." width="50%" />
<p class="caption">
Figure 10.23: K-means clustering of the data set with no structure: scatterplot of clusters for k=4. Cluster centres indicated with a cross.
</p>
</div>
</div>
</div>
<div id="example-gene-expression-profiling-of-human-tissues-1" class="section level3">
<h3><span class="header-section-number">10.4.5</span> Example: gene expression profiling of human tissues</h3>
<p>Let’s return to the data on gene expression of human tissues. Load data</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&quot;data/tissues_gene_expression/tissuesGeneExpression.rda&quot;</span>)</code></pre></div>
<p>As we saw earlier, the data set contains expression levels for over 22,000 transcripts in seven tissues.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(tissue)</code></pre></div>
<pre><code>## tissue
##  cerebellum       colon endometrium hippocampus      kidney       liver 
##          38          34          15          31          39          26 
##    placenta 
##           6</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(e)</code></pre></div>
<pre><code>## [1] 22215   189</code></pre>
<p>First we will examine the total intra-cluster variance with different values of <em>k</em>. In practice we would set <strong>nstart</strong> to a large value (e.g. 50), but in the interests of speed for this demonstration we will set it to one. We use <strong>set.seed</strong> to make this example reproducible, but in practice you would allow <strong>R</strong> to generate a random seed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">k&lt;-<span class="dv">1</span>:<span class="dv">15</span>
<span class="kw">set.seed</span>(<span class="dv">42</span>)
res_k_15 &lt;-<span class="st"> </span><span class="kw">lapply</span>(k, function(i){<span class="kw">kmeans</span>(<span class="kw">t</span>(e), i, <span class="dt">nstart=</span><span class="dv">1</span>)})
<span class="kw">plot_tot_withinss</span>(res_k_15)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:tissueExpressionElbow"></span>
<img src="09-clustering_files/figure-html/tissueExpressionElbow-1.png" alt="K-means clustering of human tissue gene expression: variance within clusters." width="100%" />
<p class="caption">
Figure 10.24: K-means clustering of human tissue gene expression: variance within clusters.
</p>
</div>
<p>If we had set <strong>nstart</strong> to a higher value we would have obtained a smoother curve in figure <a href="clustering.html#fig:tissueExpressionElbow">10.24</a>. There is no obvious elbow, but the rate of decrease in the total-within sum of squares appears to slow after k=5. Since we know that there are seven tissues in the data set we will try k=7.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">42</span>)
res &lt;-<span class="st"> </span><span class="kw">kmeans</span>(<span class="kw">t</span>(e), <span class="dv">7</span>, <span class="dt">nstart=</span><span class="dv">10</span>)
<span class="kw">table</span>(tissue, res$cluster)</code></pre></div>
<pre><code>##              
## tissue         1  2  3  4  5  6  7
##   cerebellum   0  0  0 33  0  0  5
##   colon        0  0  0  0  0 34  0
##   endometrium  0  0  0  0 15  0  0
##   hippocampus  0  0  0  0  0  0 31
##   kidney       0  0 39  0  0  0  0
##   liver       26  0  0  0  0  0  0
##   placenta     0  6  0  0  0  0  0</code></pre>
<p>The analysis has found a distinct cluster for each tissue and therefore performed slightly better than the earlier hierarchical clustering analysis, which placed endometrium and kidney observations in the same cluster.</p>
<p>To visualize the result in a 2D scatter plot we first need to apply dimensionality reduction. We will use principal component analysis (PCA), which was described in chapter <a href="dimensionality-reduction.html#dimensionality-reduction">9</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pca &lt;-<span class="st"> </span><span class="kw">prcomp</span>(<span class="kw">t</span>(e))
<span class="kw">ggplot</span>(<span class="dt">data=</span><span class="kw">as.data.frame</span>(pca$x), <span class="kw">aes</span>(PC1,PC2)) +<span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">col=</span><span class="kw">brewer.pal</span>(<span class="dv">7</span>,<span class="st">&quot;Dark2&quot;</span>)[res$cluster], 
             <span class="dt">shape=</span><span class="kw">c</span>(<span class="dv">49</span>:<span class="dv">55</span>)[res$cluster], <span class="dt">size=</span><span class="dv">5</span>) +<span class="st"> </span>
<span class="st">  </span><span class="kw">theme_bw</span>()</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:tissueExpressionPCA"></span>
<img src="09-clustering_files/figure-html/tissueExpressionPCA-1.png" alt="K-means clustering of human gene expression (k=7): scatterplot of first two principal components" width="50%" />
<p class="caption">
Figure 10.25: K-means clustering of human gene expression (k=7): scatterplot of first two principal components
</p>
</div>
</div>
</div>
<div id="dbscan" class="section level2">
<h2><span class="header-section-number">10.5</span> DBSCAN</h2>
<p>Density-based spatial clustering of applications with noise</p>
<div id="algorithm-1" class="section level3">
<h3><span class="header-section-number">10.5.1</span> Algorithm</h3>
<p>Abstract DBSCAN algorithm in pseudocode <span class="citation">(Schubert et al. <a href="#ref-Schubert2017">2017</a>)</span></p>
<pre><code>1 Compute neighbours of each point and identify core points   // Identify core points
2 Join neighbouring core points into clusters                 // Assign core points
3 foreach non-core point do
      Add to a neighbouring core point if possible            // Assign border points
      Otherwise, add to noise                                 // Assign noise points</code></pre>
<div class="figure" style="text-align: center"><span id="fig:dbscanIllustration"></span>
<img src="images/DBSCAN_Illustration.svg" alt="Illustration of the DBSCAN algorithm." width="75%" />
<p class="caption">
Figure 10.26: Illustration of the DBSCAN algorithm.
</p>
</div>
</div>
<div id="choosing-parameters" class="section level3">
<h3><span class="header-section-number">10.5.2</span> Choosing parameters</h3>
</div>
<div id="example-clustering-synthetic-data-sets-2" class="section level3">
<h3><span class="header-section-number">10.5.3</span> Example: clustering synthetic data sets</h3>
</div>
<div id="gene-expression" class="section level3">
<h3><span class="header-section-number">10.5.4</span> Gene expression</h3>
<p>tissue types?</p>
</div>
</div>
<div id="summary" class="section level2">
<h2><span class="header-section-number">10.6</span> Summary</h2>
<div id="applications" class="section level3">
<h3><span class="header-section-number">10.6.1</span> Applications</h3>
</div>
<div id="strengths" class="section level3">
<h3><span class="header-section-number">10.6.2</span> Strengths</h3>
</div>
<div id="limitations" class="section level3">
<h3><span class="header-section-number">10.6.3</span> Limitations</h3>
<!--
Not appropriate for phylogenetic analysis!!
-->
</div>
</div>
<div id="exercises-7" class="section level2">
<h2><span class="header-section-number">10.7</span> Exercises</h2>
<!--
1. Toy/synthetic clusters
2. mouse mammary time-course (kmeans and dbscan)
3. dimensionality reduction before clustering (helpful for visualization if using a partitioning method) - possibly use parasite data?
4. exercise involving heatmap

-->
<p>Exercise solutions: <a href="solutions-clustering.html#solutions-clustering">I</a></p>
<p>Solutions to exercises can be found in appendix <a href="solutions-clustering.html#solutions-clustering">I</a>.</p>

</div>
</div>



<h3>References</h3>
<div id="refs" class="references">
<div id="ref-Gionis2007">
<p>Gionis, A., H. Mannila, and P. Tsaparas. 2007. “Clustering aggregation.” <em>ACM Transactions on Knowledge Discovery from Data (TKDD)</em> 1 (1): 1–30.</p>
</div>
<div id="ref-Schubert2017">
<p>Schubert, Erich, Jörg Sander, Martin Ester, Hans Peter Kriegel, and Xiaowei Xu. 2017. “DBSCAN Revisited, Revisited: Why and How You Should (Still) Use Dbscan.” <em>ACM Trans. Database Syst.</em> 42 (3): 19:1–19:21. doi:<a href="https://doi.org/10.1145/3068335">10.1145/3068335</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="dimensionality-reduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="resources.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/bioinformatics-training/intro-machine-learning/edit/master/09-clustering.Rmd",
"text": "Edit"
},
"download": ["intro-machine-learning.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
